packages/experimental/Overflow/src/Overflow/useOverflow.ts

```diff
@@ -0,0 +1,178 @@
+      overflowItemUpdateCallbacks[id] = onItemChange;
+    },
+    [overflowItemUpdateCallbacks],
+  );
+  const disconnect = React.useCallback((id: string) => delete overflowItemUpdateCallbacks[id], [overflowItemUpdateCallbacks]);
+
+  const userSetLayoutState = React.useCallback((data: SetLayoutStateParam) => {
+    if (data.type === 'menu') {
+      setLayoutState((prev) => ({ ...prev, menu: data.layoutDone }));
+    } else {
+      setLayoutState((prev) => ({ ...prev, items: { ...prev.items, [data.id]: data.layoutDone } }));
+    }
+  }, []);
+
+  const onContainerLayout = React.useCallback(
+    (e: LayoutChangeEvent) => {
+      const { width, height } = e.nativeEvent.layout;
+      if (width !== undefined && height !== undefined) {
+        setContainerSize({ width, height });
+      }
+      onLayout && onLayout(e);
+    },
+    [onLayout],
+  );
+
+  const handleItemUpdate = React.useCallback(
+    (item: OverflowItemEntry) => {
+      if (overflowManager.hasItem(item.id)) {
+        overflowManager.updateItem(item.id, item);
+      } else {
+        overflowManager.addItem(item);
+      }
+    },
+    [overflowManager],
+  );
+
+  const onMenuLayout = React.useCallback((size: LayoutSize) => overflowManager.setMenuSize(size), [overflowManager]);
+
+  // Callback for when the overflow manager has calculated all overall changes towards item visibilities in its update cycle.
+  const onOverflowUpdate = React.useCallback(
+    (data: OverflowUpdatePayload) => {
+      setOverflowState((prev) => {
+        const visibilities: Record<string, boolean> = {};
+        for (const id of data.visibleIds) {
+          visibilities[id] = true;
+        }
+        for (const id of data.invisibleIds) {
+          visibilities[id] = false;
+        }
+        return { ...prev, itemVisibility: visibilities, hasOverflow: data.invisibleIds.length > 0 };
+      });
+      overflowUpdateCallback && overflowUpdateCallback(data);
+    },
+    [overflowUpdateCallback],
+  );
+
+  // If there is one item showing, callback to send the dimensions the item should have to truncate itself and have everything fit neatly
+  // in the overflow container.
+  const onUpdateItemDimension = React.useCallback(
+    (data: ItemDimensionUpdatePayload) => {
+      overflowItemUpdateCallbacks[data.id] && overflowItemUpdateCallbacks[data.id]({ type: 'layout', id: data.id, newLayout: data.update });
+    },
+    [overflowItemUpdateCallbacks],
+  );
+
+  // Callback for when a single item has its visibility changed.
+  const onUpdateItemVisibility = React.useCallback(
+    (data: ItemVisibilityUpdatePayload) => {
+      overflowItemUpdateCallbacks[data.id] && overflowItemUpdateCallbacks[data.id]({ type: 'visibility', ...data });
+    },
+    [overflowItemUpdateCallbacks],
+  );
+
+  // This layout effect is run before re-rendering, allowing us to calculate what's visible / invisible with our overflow manager.
+  React.useLayoutEffect(() => {
+    if (containerSize) {
```

Comment:
super nit: This is a personal styling thing, but I think it's easier to read code if there's less nested cases. If code should only run if a condition is satisfied, you can have it return early if that condition isn't satisfied instead.

instead of 
```
if (foo) {
    if (bar) {
    } else {
    }
}
```

I would write
```
if (!foo)
    return;

if (bar) {
} else {
}
```